Source/NSOrderedSet.m

```diff
@@ -0,0 +1,1894 @@
+{
+  [self subclassResponsibility: _cmd];
+}
+
+- (void)setObject:(id)object atIndexedSubscript:(NSUInteger)index
+{
+  if ([self count] == index)
+    {
+      [self addObject: object];
+    }
+  else
+    {
+      [self replaceObjectAtIndex: index withObject: object];
+    }
+}
+
+- (void)insertObjects:(NSArray *)array atIndexes:(NSIndexSet *)indexes
+{
+  NSUInteger	index = [indexes firstIndex];
+  NSEnumerator	*enumerator = [array objectEnumerator];
+  id		object = [enumerator nextObject];
+ 
+  while (object != nil && index != NSNotFound)
+    {
+      [self insertObject: object atIndex: index];
+      object = [enumerator nextObject];
+      index = [indexes indexGreaterThanIndex: index];
+    }
+}
+
+- (void)removeObject:(id)anObject
+{
+  NSUInteger	i;
+
+  if (anObject == nil)
+    {
+      NSWarnMLog(@"attempt to remove nil object");
+      return;
+    }
+  i = [self count];
+  if (i > 0)
+    {
+      IMP	rem = 0;
+      IMP	get = [self methodForSelector: oaiSel];
+      BOOL	(*eq)(id, SEL, id)
+	= (BOOL (*)(id, SEL, id))[anObject methodForSelector: eqSel];
+
+      while (i-- > 0)
+	{
+	  id	o = (*get)(self, oaiSel, i);
+
+	  if (o == anObject || (*eq)(anObject, eqSel, o) == YES)
+	    {
+	      if (rem == 0)
+		{
+		  rem = [self methodForSelector: remSel];
+		  /*
+		   * We need to retain the object so that when we remove the
+		   * first equal object we don't get left with a bad object
+		   * pointer for later comparisons.
+		   */
+		  // RETAIN(anObject);
+		}
+	      (*rem)(self, remSel, i);
+	      break;  // since this is a set we should only have one copy...
+	    }
+	}
+      //if (rem != 0)
+      // {
+      //  RELEASE(anObject);
+      // }
+    }
+}
+
+- (void)removeObjectAtIndex:(NSUInteger)index  // required override
+{
+  [self subclassResponsibility: _cmd];
+}
+
+- (void) _removeObjectsFromIndices: (NSUInteger*)indices
+			numIndices: (NSUInteger)count
+{
+  if (count > 0)
+    {
+      NSUInteger	to = 0;
+      NSUInteger	from = 0;
+      NSUInteger	i;
+      GS_BEGINITEMBUF(sorted, count, NSUInteger);
+
+      while (from < count)
+	{
+	  NSUInteger	val = indices[from++];
+
+	  i = to;
+	  while (i > 0 && sorted[i-1] > val)
+	    {
+	      i--;
+	    }
+	  if (i == to)
+	    {
+	      sorted[to++] = val;
+	    }
+	  else if (sorted[i] != val)
+	    {
+	      NSUInteger	j = to++;
+
+	      if (sorted[i] < val)
+		{
+		  i++;
+		}
+	      while (j > i)
+		{
+		  sorted[j] = sorted[j-1];
+		  j--;
+		}
+	      sorted[i] = val;
+	    }
+	}
+
+      if (to > 0)
+	{
+	  IMP	rem = [self methodForSelector: remSel];
+
+	  while (to--)
+	    {
+	      (*rem)(self, remSel, sorted[to]);
+	    }
+	}
+      GS_ENDITEMBUF();
+    }
+}
+
+- (void)removeObjectsAtIndexes:(NSIndexSet *)indexes
+{
+  NSUInteger count = [indexes count];
+  NSUInteger indexArray[count];
+
+  [indexes getIndexes: indexArray
+             maxCount: count
+         inIndexRange: NULL];
+
+  [self _removeObjectsFromIndices: indexArray
```

Comment:
I am sure I already wrote a comment that you don't have to sort the indexes here but somehow this got lost.

Reply:
I closed it because this method is actually taken from NSArray.  I'm curious though, how do we know the indexes are sorted?

Reply:
I mean.. we DO in that particular method, but those indexes could come from another source.

Reply:
Simplified this method as discussed.